# Part 1: Forward Mode Automatic Differentiation

Forward mode AD can simply be implemented by defining a class to represent [dual numbers](https://en.wikipedia.org/wiki/Dual_number) which hold the value and its derivative. The following skeleton defines a dual number and implements multiplication. 

__Tasks:__

- Addition (`__add__`) is incomplete - can you finish it? 
- Can you also implement division (`__truediv__`), subtraction (`__sub__`) and power (`__pow__`)?

In [32]:
import math

class DualNumber:
    def __init__(self, value, dvalue):
        self.value = value
        self.dvalue = dvalue

    def __str__(self):
        return str(self.value) + " + " + str(self.dvalue) + "ε"

    def __mul__(self, other):
        return DualNumber(self.value * other.value,
            self.dvalue * other.value + other.dvalue * self.value)
    
    def __add__(self, other):
        return DualNumber(self.value + other.value,
            self.dvalue + other.dvalue)
    
    def __truediv__(self, other):
        value = self.value / other.value
        dvalue = self.dvalue * other.value + ((self.dvalue * other.value - self.value * other.dvalue) / (other.value ** 2))
        return DualNumber(value, dvalue)    
    
    def __sub__(self, other):
        return DualNumber(self.value - other.value,
            self.dvalue - other.dvalue)     

    def __pow__(self, other):
        value = self.value ** other.value     
        dvalue = value * (other.dvalue * math.log(self.value) + self.dvalue * other.value / self.value)
        return DualNumber(value, dvalue)


In [34]:
# Tests

dn = DualNumber(1,0) + DualNumber(1,0) / DualNumber(1,0) - DualNumber(1,0)**DualNumber(1,0)
print(dn)

1.0 + 0.0ε


## Implementing math functions

We also need to implement some core math functions. Here's the sine function for a dual number:

In [36]:
def sin(x):
    return DualNumber(math.sin(x.value), math.cos(x.value)*x.dvalue)

__Task:__ can you implement the _cosine_ (`cos`), _tangent_ (`tan`), and _exponential_ (`exp`) functions in the code block below?

In [37]:
# TODO: implement additional math functions on dual numbers

def cos(x):
    return DualNumber(math.cos(x.value), - math.sin(x.value)*x.dvalue)

def tan(x):
    return DualNumber(math.tan(x.value), x.dvalue / (math.cos(x.value) ** 2))

def exp(x):
    return DualNumber(math.exp(x.value), x.dvalue * math.exp(x.value))

In [38]:
# Tests
assert cos(DualNumber(0,0)).value == 1
assert tan(DualNumber(0,0)).value == 0
assert exp(DualNumber(0,0)).value == 1


## Time to try it out

We're now in a position to try our implementation.

__Task:__ 

- Try running the following code to compute the value of the function $z=x\cdot y+sin(x)$ given $x=0.5$ and $y=4.2$, together with the derivative $\partial z/\partial x$ at that point. 

In [42]:
dnx = DualNumber(0.5, 1)
dny = DualNumber(4.2, 0)
z = dnx * dny + sin(dnx)

2.579425538604203 + 5.077582561890373ε


__Task__: Differentiate the above function with respect to $x$ and write the symbolic derivatives in the following box. Verify the result computed above is correct by plugging-in the values into your symbolic gradient expression.

$\frac{\delta z}{\delta x}=y+cos(x)=4.2 + cos(0.5)=5.0776$ 

__Task:__ Now use the code block below to compute the derivative $\partial z/\partial y$ of the above expression (at the same point $x=0.5, y=4.2$ as above) and store the derivative in the variable `dzdy` (just the derivative, not the Dual Number). Verify by hand that the result is correct.

In [45]:
dnx = DualNumber(0.5, 0)
dny = DualNumber(4.2, 1)
z = dnx * dny + sin(dnx)
dzdy = z.dvalue
print('dz/dy:', dzdy)
assert(dzdy == dnx.value)

dz/dy: 0.5


In [44]:
#Tests
assert dzdy
assert type(dzdy) == float


__Task:__ Finally, use the code block below to experiment and test the other math functions and methods you created.

In [ ]:
# YOUR CODE HERE
raise NotImplementedError()